### PR TITLE
Git Gateway support for BitBucket

### PR DESCRIPTION
**Pull request by Benaiah**
**- Summary**

Adds BitBucket support to Git Gateway, including logic to refresh access tokens using a provided refresh token, OAuth client ID, and OAuth client secret.

~This PR needs link rewriting added before it's merged.~ Link rewriting has been added.

**- Test plan**

Tested manually.

**- Description for the changelog**

Git gateway support for BitBucket

### CODE REVIEW

File: api/bitbucket.go [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+
+	return errors.New("Access to endpoint not allowed: your role doesn't allow access")
+}
+
+func rewriteBitBucketLink(link, endpointAPIURL, proxyAPIURL string) string {
+	return proxyAPIURL + strings.TrimPrefix(link, endpointAPIURL)
+}
+
+func rewriteLinksInBitBucketResponse(resp *http.Response, endpointAPIURL, proxyAPIURL string) error {
+	body, err := ioutil.ReadAll(resp.Body)
```
Comment (Benaiah) [inline api/bitbucket.go]: This needs to correctly handle gzipped content - it currently causes an error if you request gzipped content.

File: api/bitbucket.go [hunk 2]
```diff
@@ -178,14 +180,26 @@ func rewriteBitBucketLink(link, endpointAPIURL, proxyAPIURL string) string {
 }
 
 func rewriteLinksInBitBucketResponse(resp *http.Response, endpointAPIURL, proxyAPIURL string) error {
-	body, err := ioutil.ReadAll(resp.Body)
+	var bodyReader io.ReadCloser
+	var err error
+	switch resp.Header.Get("Content-Encoding") {
+	case "gzip":
+		bodyReader, err = gzip.NewReader(resp.Body)
+		defer bodyReader.Close()
```
Comment (calavera) [inline api/bitbucket.go]: the `defer` probably needs to go after checking the error.

File: api/bitbucket.go [hunk 3]
```diff
@@ -196,9 +210,21 @@ func rewriteLinksInBitBucketResponse(resp *http.Response, endpointAPIURL, proxyA
...
+
+	switch resp.Header.Get("Content-Encoding") {
+	case "gzip":
+		var compressedBody bytes.Buffer
+		w := gzip.NewWriter(&compressedBody)
+		_, err = w.Write(newBodyBytes)
+		if err != nil {
+			return err
+		}
+		w.Close()
```
Comment (calavera) [inline api/bitbucket.go]: this should go after `w := szip.NewWriter` with a `defer`

File: api/bitbucket.go [hunk 4]
```diff
@@ -169,13 +173,57 @@ func (bb *BitBucketGateway) authenticate(w http.ResponseWriter, r *http.Request)
...
+	if next, ok := b["next"].(string); ok {
+		b["next"] = rewriteBitBucketLink(next, endpointAPIURL, proxyAPIURL)
+	}
+
+	if prev, ok := b["previous"].(string); ok {
+		b["previous"] = rewriteBitBucketLink(prev, endpointAPIURL, proxyAPIURL)
+	}
+
+	newBody, err := json.Marshal(b)
+	println(newBody)
```
Comment (calavera) [inline api/bitbucket.go]: remove this.

**Comment by Benaiah** [comment]
@calavera updated with requested changes.

**Comment by amarpathak** [comment]
So do we have lit gateway support with bitbuckket ?

**Pull request merged**